### DISCUSSION TITLE
Transfering a large file over Thread (~250KB)

### DISCUSSION THREAD

**Discussion by dakhnod** [openthread/openthread#11018] [Q&A]
We need to transfer firmware files over a large Thread network.
While I saw https://github.com/openthread/openthread/issues/5498, and this would pose a viable solution, I wanted to ask if there is a better solution for transferring bigger blocks of data over a Thread network.

What would be the best approach to send files sized around 250KB over Thread?

**Comment by jwhui** [comment]
Consider using TCP - OpenThread includes a TCP implementation.

https://github.com/openthread/openthread/blob/main/src/cli/README_TCP.md

**Reply by dakhnod** [reply]
Would you consider TCP a good solution, considering the specifics of Thread? 
Could something like Q-Block CoAP be more beneficial?

**Reply by dakhnod** [reply]
Also, what what speed would you expect for a 256kB transfer? To do it via CoAP/HTTP it takes me about 30secs. Is that realistic?

**Reply by jwhui** [reply]
A primary benefit of TCP is that it supports a robust congestion-control mechanism that is widely deployed, well understood, and has stood the test of time.

The speed of transfer is largely dependent on environmental conditions, including network topology, external interference, and other concurrent network activity.

For reference, check out this [NSDI'20 paper](https://www.usenix.org/conference/nsdi20/presentation/kumar) on TCP in Thread networks. I would be interested to any real-world evaluation / comparison of Q-Block CoAP.

**Comment by dakhnod** [comment]
Thanks, you're of really great help!

What is the protocol overhead of spinel?

Using a (real) UART baudrate of 115200, what net network bandwidth can we expect?

Also, what bandwidth can we expect using the nordic RCP, without real UART in the way?

**Reply by jwhui** [reply]
Goodput numbers are published in the NSDI'20 paper linked above:

- 75 kbps for one-hop
- 20 kbps for multi-hop

For UART baudrate, I recommend increasing to 460800.

**Reply by abtink** [reply]
Generally, in an RCP/host architecture (intended for deployment), it is good practice to ensure the Spinel bus is fast enough that it is not the bottleneck for data transfer.
